src/NUnitFramework/framework/Constraints/Comparers/DictionariesComparer.cs

```diff
@@ -54,8 +54,12 @@ internal DictionariesComparer(NUnitEqualityComparer equalityComparer)
                 return false;
 
             foreach (object key in xDictionary.Keys)
-                if (!_equalityComparer.AreEqual(xDictionary[key], yDictionary[key], ref tolerance, topLevelComparison: false))
+            {
+                state.TopLevelComparison = false;
+                if (!_equalityComparer.AreEqual(xDictionary[key], yDictionary[key], ref tolerance, state))
```

Comment:
I find it very much simpler to think about when the state type is readonly because the syntax then shows that shared state is not being mutated, e.g. `state.WithTopLevelComparison(false)`.

(Shared state isn't being mutated here because `state` is a value type and `TopLevelComparison` is backed by a value-typed field, but I'm still happier when the syntax makes it obvious.)